### PR TITLE
feat(generator): optimize round-robin decorators

### PR DESCRIPTION
**Pull request by coryan**
Thanks to @devbww for the suggestion

This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10093)

### CODE REVIEW

File: generator/internal/round_robin_decorator_generator.cc [hunk 1]
```diff
@@ -372,9 +372,10 @@ future<Status>
   CcPrint(R"""(
 std::shared_ptr<$stub_class_name$>
 $round_robin_class_name$::Child() {
-  std::lock_guard<std::mutex> lk(mu_);
-  auto child = children_[current_];
+  std::unique_lock<std::mutex> lk(mu_);
+  auto const& child = children_[current_];
```
Comment (devbww) [inline generator/internal/round_robin_decorator_generator.cc:376]: And perhaps `auto const current = current_;` and then `return children_[current];` would move some more arithmetic outside the lock.

File: generator/internal/round_robin_decorator_generator.cc [hunk 2]
```diff
@@ -372,9 +372,10 @@ future<Status>
...
   current_ = (current_ + 1) % children_.size();
```
Comment (devbww) [inline generator/internal/round_robin_decorator_generator.cc:377]: You might also consider strength reduction to ...
```
  if (++current_ == children_.size()) current_ = 0;
```

**Review by devbww — APPROVED** [review]
Some more nits on the nit, if they're convincing.

File: generator/internal/round_robin_decorator_generator.cc (hunk 1 above)
Comment (coryan) [reply generator/internal/round_robin_decorator_generator.cc:376]: I think those would make no difference with modern compilers:

https://godbolt.org/z/vaa5Yq3hr

But if you think they can make the intent better, I would be happy to use that.

File: generator/internal/round_robin_decorator_generator.cc (hunk 2 above)
Comment (coryan) [reply generator/internal/round_robin_decorator_generator.cc:377]: Ditto

Comment (coryan) [reply generator/internal/round_robin_decorator_generator.cc]: I am going to go with your approach.  easier to read I think.

**Pull request merged**